elasticsearch_dsl/_async/index.py

```diff
@@ -0,0 +1,540 @@
+#  Licensed to Elasticsearch B.V. under one or more contributor
+#  license agreements. See the NOTICE file distributed with
+#  this work for additional information regarding copyright
+#  ownership. Elasticsearch B.V. licenses this file to you under
+#  the Apache License, Version 2.0 (the "License"); you may
+#  not use this file except in compliance with the License.
+#  You may obtain a copy of the License at
+#
+# 	http://www.apache.org/licenses/LICENSE-2.0
+#
+#  Unless required by applicable law or agreed to in writing,
+#  software distributed under the License is distributed on an
+#  "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+#  KIND, either express or implied.  See the License for the
+#  specific language governing permissions and limitations
+#  under the License.
+
+from ..async_connections import get_connection
+from ..exceptions import IllegalOperation
+from ..index_base import IndexBase
+from .mapping import AsyncMapping
+from .search import AsyncSearch
+from .update_by_query import AsyncUpdateByQuery
+
+
+class AsyncIndexTemplate:
+    def __init__(self, name, template, index=None, order=None, **kwargs):
+        if index is None:
+            self._index = AsyncIndex(template, **kwargs)
+        else:
+            if kwargs:
+                raise ValueError(
+                    "You cannot specify options for Index when"
+                    " passing an Index instance."
+                )
+            self._index = index.clone()
+            self._index._name = template
+        self._template_name = name
+        self.order = order
+
+    def __getattr__(self, attr_name):
+        return getattr(self._index, attr_name)
+
+    def to_dict(self):
+        d = self._index.to_dict()
+        d["index_patterns"] = [self._index._name]
+        if self.order is not None:
+            d["order"] = self.order
+        return d
+
+    async def save(self, using=None):
+        es = get_connection(using or self._index._using)
+        return await es.indices.put_template(
+            name=self._template_name, body=self.to_dict()
+        )
+
+
+class AsyncIndex(IndexBase):
+    def _get_connection(self, using=None):
+        if self._name is None:
+            raise ValueError("You cannot perform API calls on the default index.")
+        return get_connection(using or self._using)
+
+    connection = property(_get_connection)
+
+    def get_or_create_mapping(self):
+        if self._mapping is None:
+            self._mapping = AsyncMapping()
+        return self._mapping
```

Comment:
I was confused for a bit because this sync method in `AsyncIndex` is only called by `IndexBase`.

Is there value in making `IndexBase` (and other bases) as inheriting from [`abc.ABC`](https://docs.python.org/3/library/abc.html) to avoid direct instantiation and declaring sync methods that are in both async/sync child classes as abstract methods in `IndexBase`? That way `get_or_create_mapping` *will* show up in `IndexBase`, but will stay implemented in `AsyncIndex` and `Index`. The same applies to `as_template`.

Would it be useful to do for async methods too? Not sure. Especially as the base will have to be either async or sync, which will require type ignores in the child classes when we add mypy.

Reply:
The `get_or_create_mapping()` method is also called by the `Index` and `AsyncIndex` classes, in the `load_mapping()` method:

```python
async def load_mappings(self, using=None):
        await self.get_or_create_mapping().update_from_es(
            self._name, using=using or self._using
        )
```

But given that this method is sync in both interfaces, I think for this one we should move the method to the base class, then add a `get_mapping_class()` in the subclasses each returning its variant of the mapping class. You think?

Reply:
And also, maybe we should make all these base classes abstract, but honestly I have never worried about that in other projects of my own, it always seemed like have a `Base` prefix or suffix was sufficient for people to know this isn't a class meant to be instantiated.

Reply:
Sure, that works, and is better, but it only moves the problem. I continue to think that:

 * making the Base classes inheriting from `abc.ABC` is useful, at least to prevent direct instantiation
 * declaring sync methods from child classes (`get_mapping_class()` now) is probably helpful?

Reply:
Our comments crossed. I'm happy to leave out `ABC` and abstract methods if you don't think they help.